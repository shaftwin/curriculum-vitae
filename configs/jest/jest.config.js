const path = require('path');

module.exports = {
  rootDir: '../../',
  testEnvironment: 'jsdom',
  testMatch: ['**/*.test.js'],
  setupFilesAfterEnv: ['./node_modules/@testing-library/jest-dom'],
  moduleDirectories: [path.resolve(__dirname, '../../'), 'node_modules'],
  setupFiles: [
    './configs/jest/mock.js',
    './configs/cucumber/jest-cucumber-config.js',
  ],
  snapshotSerializers: ['@emotion/jest/serializer'],
};
