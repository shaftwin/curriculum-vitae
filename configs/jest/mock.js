jest.mock('react-i18next', () => ({
  withTranslation: () => (Component) => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => '' };
    return Component;
  },
  // this mock makes sure any components using the
  // translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));
