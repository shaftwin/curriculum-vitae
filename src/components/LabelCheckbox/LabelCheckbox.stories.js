import React from 'react';
import LabelCheckbox from './index';

export default {
  title: 'Components/LabelCheckBox',
  component: LabelCheckbox,
};

// eslint-disable-next-line react/jsx-props-no-spreading
const Template = (args) => <LabelCheckbox {...args} />;

export const Default = Template.bind({});

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Label',
};

export const WithError = Template.bind({});
WithError.args = {
  label: 'Label',
  error: true,
  errorMessage: 'Error message',
};

export const WithCustomComponent = Template.bind({});
WithCustomComponent.args = {
  label: (
    <div>
      <label>test label </label>
      <a href="#">my test link</a>
    </div>
  ),
};
