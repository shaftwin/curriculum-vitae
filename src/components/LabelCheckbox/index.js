/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import PropTypes from 'prop-types';
import useLabelCheckboxSetup from './hooks/useLabelCheckboxSetup';
import {
  Container,
  CheckboxContainer,
  StyledCheckbox,
  HiddenCheckbox,
  CustomCheckboxIcon,
  ErrorMessage,
} from './styled';

const LabelCheckbox = ({
  label,
  labelPosition,
  value,
  error,
  size,
  disabled,
  className,
  errorMessage,
  ...props
}) => {
  const { state, checked, labelPart, setHover, handleChange } =
    useLabelCheckboxSetup({
      disabled,
      error,
      value,
      label,
      labelPosition,
      size,
    });

  return (
    <Container
      state={state}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      data-testid="LabelCheckboxContainer"
      className={className}
    >
      <CheckboxContainer
        data-testid="checkboxContainer"
        size={size}
        labelPosition={labelPosition}
      >
        {labelPart}
        <HiddenCheckbox
          data-testid="hiddenCheckbox"
          type="checkbox"
          checked={checked}
          onChange={handleChange}
          disabled={disabled}
          {...props}
        />
        <StyledCheckbox
          data-testid="styledCheckbox"
          state={state}
          checked={checked}
        >
          {checked && <CustomCheckboxIcon state={state} />}
        </StyledCheckbox>
      </CheckboxContainer>
      {state === 'error' && !!errorMessage && (
        <ErrorMessage data-testid="errorMessage">{errorMessage}</ErrorMessage>
      )}
    </Container>
  );
};

LabelCheckbox.propTypes = {
  /**
   * Optional label
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  /**
   * Optional labelPosition
   */
  labelPosition: PropTypes.oneOf(['left', 'right']),
  /**
   * Optional value
   */
  value: PropTypes.bool,
  /**
   * Optional errorMessage
   */
  error: PropTypes.bool,
  /**
   * Optional error
   */
  errorMessage: PropTypes.string,
  /**
   * Optional disabled
   */
  disabled: PropTypes.bool,
  /**
   * Optional size
   */
  size: PropTypes.oneOf(['small', 'normal']),
  /**
   * Optional custom style
   */
  className: PropTypes.string,
};

LabelCheckbox.defaultProps = {
  label: null,
  labelPosition: 'left',
  value: false,
  error: false,
  errorMessage: null,
  disabled: false,
  size: 'normal',
};

export default LabelCheckbox;
