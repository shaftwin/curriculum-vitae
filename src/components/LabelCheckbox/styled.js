import styled from '@emotion/styled';
import CheckboxIcon from '../../../assets/svg/checkboxIcon';

/*
 In this file, the hover properties of the components are all based on the hover
event of the main Container
*/

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  color: ${({ theme, state }) => theme.labelCheckbox.textColor[state]};
`;

export const CheckboxContainer = styled.label`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: ${({ labelPosition }) =>
    labelPosition === 'left' ? 'row' : 'row-reverse'};
  margin-bottom: 0.375rem;
`;

export const CustomCheckboxIcon = styled(CheckboxIcon)`
  fill: ${({ theme, state }) => theme.labelCheckbox.iconColor[state]};
`;

export const HiddenCheckbox = styled.input`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
`;

export const StyledCheckbox = styled.div`
  width: 16px;
  height: 16px;
  display: flex;
  justify-content: center;
  align-items: center;
  background: ${({ theme, state, checked }) =>
    checked
      ? theme.labelCheckbox.background.selected[state]
      : theme.labelCheckbox.background.unselected[state]};
  border-radius: 0.25rem;
  border: ${({ theme, state, checked }) =>
    checked ? '0' : theme.labelCheckbox.border.unselected[state]};
  cursor: ${({ theme, state }) => theme.global.cursor[state]};
`;

export const LabelContainer = styled.div`
  margin: ${({ theme, labelPosition }) =>
    theme.labelCheckbox.labelMargin[labelPosition]};
`;

export const CustomText = styled.label`
  color: ${({ theme, state }) => theme.labelCheckbox.textColor[state]};
  font-size: ${({ theme, size }) => theme.labelCheckbox.fontSize[size]};
`;

export const ErrorMessage = styled.label`
  color: ${({ theme }) => theme.labelCheckbox.errorMessage.color};
  font-size: ${({ theme }) => theme.labelCheckbox.errorMessage.fontSize};
  font-weight: ${({ theme }) => theme.labelCheckbox.errorMessage.fontWeight};
`;
