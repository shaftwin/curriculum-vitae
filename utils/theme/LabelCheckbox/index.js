import { global } from '../global';

export const labelCheckbox = {
  iconColor: {
    default: global.colors.white,
    hover: global.colors.hover.white,
    disabled: global.colors.disabledGrey,
    error: global.colors.white,
  },
  textColor: {
    default: global.colors.darkBlue,
    hover: global.colors.hover.darkBlue,
    disabled: global.colors.disabledGrey,
    error: global.colors.darkBlue,
  },
  fontSize: {
    small: global.fontSize[11],
    normal: global.fontSize[13],
  },
  labelMargin: {
    // The margin is the opposite of the labelPosition
    left: '0rem 0.625rem 0rem 0rem',
    right: '0rem 0rem 0rem 0.625rem',
  },
  background: {
    unselected: {
      default: global.colors.white,
      hover: global.colors.white,
      disabled: global.colors.grey,
      error: global.colors.white,
    },
    selected: {
      default: global.colors.green,
      hover: global.colors.hover.green,
      disabled: global.colors.grey,
      error: global.colors.status.error,
    },
  },
  border: {
    unselected: {
      default: `0.063rem solid ${global.colors.mediumGrey}`,
      hover: `0.063rem solid ${global.colors.brownGrey}`,
      error: `0.063rem solid ${global.colors.status.error}`,
    },
  },
  errorMessage: {
    color: global.colors.status.error,
    fontSize: global.fontSize[10],
    fontWeight: global.fontWeight.semiBold,
  },
};
