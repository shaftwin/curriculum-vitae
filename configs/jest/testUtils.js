// import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import React from 'react';
import theme from '../../utils/theme';

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
