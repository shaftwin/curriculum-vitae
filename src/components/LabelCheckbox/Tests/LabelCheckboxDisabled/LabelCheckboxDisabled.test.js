import * as React from 'react';
import { screen } from '@testing-library/react';
import { defineFeature, loadFeature } from 'jest-cucumber';
import userEvent from '@testing-library/user-event';
import { render } from '../../../../../configs/jest/testUtils';
import LabelCheckbox from '../..';

const feature = loadFeature(
  // eslint-disable-next-line max-len
  'src/components/LabelCheckbox/Tests/LabelCheckboxDisabled/LabelCheckboxDisabled.feature',
);

defineFeature(feature, (test) => {
  let container;

  const givenLabelCheckboxIsRendered = (given, component) => {
    given('LabelCheckbox component is rendered', () => {
      container = render(component).container;
    });
  };

  test('Disabled LabelCheckbox should not allow to check', ({
    given,
    and,
    when,
    then,
  }) => {
    givenLabelCheckboxIsRendered(
      given,
      <LabelCheckbox disabled label="testLabel" />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByTestId('LabelCheckboxContainer')).toBeInTheDocument();
      expect(screen.getByTestId('checkboxContainer')).toBeInTheDocument();
      expect(screen.getByTestId('labelContainer')).toBeInTheDocument();
    });

    and('Undesirable elements are not mounted', () => {
      expect(screen.queryByTestId('checkboxIcon')).not.toBeInTheDocument();
      expect(screen.queryByTestId('errorMessage')).not.toBeInTheDocument();
    });

    and('LabelCheckbox has the disabled style', () => {
      expect(container).toMatchSnapshot();
    });

    when('Click event is fired on Checkbox', () => {
      userEvent.click(screen.getByTestId('hiddenCheckbox'));
    });

    then('The checkbox is still not checked', () => {
      expect(screen.queryByTestId('checkboxIcon')).not.toBeInTheDocument();
    });
  });

  test('Disabled LabelCheckbox with value set to true', ({ given, and }) => {
    givenLabelCheckboxIsRendered(
      given,
      <LabelCheckbox disabled value={true} label="testLabel" />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByTestId('LabelCheckboxContainer')).toBeInTheDocument();
      expect(screen.getByTestId('checkboxContainer')).toBeInTheDocument();
      expect(screen.getByTestId('labelContainer')).toBeInTheDocument();
    });

    and('Undesirable elements are not mounted', () => {
      expect(screen.queryByTestId('errorMessage')).not.toBeInTheDocument();
    });

    and('LabelCheckbox has the disabled style', () => {
      expect(container).toMatchSnapshot();
    });

    and('Checkbox is checked', () => {
      expect(screen.getByTestId('checkboxIcon')).toBeInTheDocument();
    });
  });
});
