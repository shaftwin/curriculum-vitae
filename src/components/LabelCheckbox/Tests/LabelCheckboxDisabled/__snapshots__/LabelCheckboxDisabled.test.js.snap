// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelCheckBoxDisabled Disabled LabelCheckbox should not allow to check 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  color: #888596;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 0.375rem;
}

.emotion-2 {
  margin: 0rem 0.625rem 0rem 0rem;
}

.emotion-3 {
  color: #888596;
  font-size: 0.813rem;
}

.emotion-4 {
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
}

.emotion-5 {
  width: 16px;
  height: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #EBEEF0;
  border-radius: 0.25rem;
  cursor: not-allowed;
}

<div>
  <div
    class="emotion-0"
    data-testid="LabelCheckboxContainer"
  >
    <label
      class="emotion-1"
      data-testid="checkboxContainer"
    >
      <div
        class="emotion-2"
        data-testid="labelContainer"
      >
        <label
          class="emotion-3"
        >
          testLabel
        </label>
      </div>
      <input
        class="emotion-4"
        data-testid="hiddenCheckbox"
        disabled=""
        type="checkbox"
      />
      <div
        class="emotion-5"
        data-testid="styledCheckbox"
      />
    </label>
  </div>
</div>
`;

exports[`LabelCheckBoxDisabled Disabled LabelCheckbox with value set to true 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  color: #888596;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 0.375rem;
}

.emotion-2 {
  margin: 0rem 0.625rem 0rem 0rem;
}

.emotion-3 {
  color: #888596;
  font-size: 0.813rem;
}

.emotion-4 {
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
}

.emotion-5 {
  width: 16px;
  height: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #EBEEF0;
  border-radius: 0.25rem;
  border: 0;
  cursor: not-allowed;
}

.emotion-6 {
  fill: #888596;
}

<div>
  <div
    class="emotion-0"
    data-testid="LabelCheckboxContainer"
  >
    <label
      class="emotion-1"
      data-testid="checkboxContainer"
    >
      <div
        class="emotion-2"
        data-testid="labelContainer"
      >
        <label
          class="emotion-3"
        >
          testLabel
        </label>
      </div>
      <input
        checked=""
        class="emotion-4"
        data-testid="hiddenCheckbox"
        disabled=""
        type="checkbox"
      />
      <div
        class="emotion-5"
        data-testid="styledCheckbox"
      >
        <svg
          class="emotion-6"
          data-testid="checkboxIcon"
          fill="white"
          height="7"
          viewBox="0 0 10 7"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 0C8.72 0 8.47 0.11 8.29 0.29L4 4.59L1.71 2.29C1.53 2.11 1.28 2 1 2C0.45 2 0 2.45 0 3C0 3.28 0.11 3.53 0.29 3.71L3.29 6.71C3.47 6.89 3.72 7 4 7C4.28 7 4.53 6.89 4.71 6.71L9.71 1.71C9.89 1.53 10 1.28 10 1C10 0.45 9.55 0 9 0Z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;
